File: examples/swift/flappy_bird/Sources/FlappyBird/ScoreCounter.swift
// ScoreCounter.swift

import GoudEngine

final class ScoreCounter {
    private(set) var score: Int = 0
    private var digitTextures: [UInt64] = Array(repeating: 0, count: 10)

    private var xOffset: Float = 0
    private var yOffset: Float = 50

    private let digitWidth: Float = 24
    private let digitHeight: Float = 36
    private let digitSpacing: Float = 30

    func initialize(game: GoudGame) {
        for i in 0..<10 {
            digitTextures[i] = game.loadTexture(path: "assets/sprites/\(i).png")
        }
        xOffset = Float(GameConstants.screenWidth) / 2 - 30
    }

    func incrementScore() {
        score += 1
    }

    func resetScore() {
        score = 0
    }

    func draw(game: GoudGame) {
        let scoreString = String(score)
        for (i, char) in scoreString.enumerated() {
            let digit = Int(String(char))!
            game.drawSprite(
                texture: digitTextures[digit],
                x: xOffset + Float(i) * digitSpacing + digitWidth / 2,
                y: yOffset + digitHeight / 2,
                width: digitWidth,
                height: digitHeight
            )
        }
    }
}

File: examples/kotlin/flappy_bird/src/main/kotlin/ScoreCounter.kt
import com.goudengine.core.GoudGame

class ScoreCounter {
    var score: Int = 0
        private set

    private val digitTextures = LongArray(10)

    private var xOffset = 0f
    private var yOffset = 50f

    companion object {
        const val DIGIT_WIDTH = 24f
        const val DIGIT_HEIGHT = 36f
        const val DIGIT_SPACING = 30f
    }

    fun initialize(game: GoudGame) {
        for (i in 0 until 10) {
            digitTextures[i] = game.loadTexture("assets/sprites/$i.png")
        }
        xOffset = GameConstants.SCREEN_WIDTH / 2f - 30f
    }

    fun increment() { score++ }
    fun reset() { score = 0 }

    fun draw(game: GoudGame) {
        val scoreString = score.toString()
        for ((i, ch) in scoreString.withIndex()) {
            val digit = ch - '0'
            game.drawSprite(
                digitTextures[digit],
                xOffset + i * DIGIT_SPACING + DIGIT_WIDTH / 2f,
                yOffset + DIGIT_HEIGHT / 2f,
                DIGIT_WIDTH,
                DIGIT_HEIGHT
            )
        }
    }
}
